### ISSUE TITLE
[ActiveRecord] Relationship methods accept a join clause

### ISSUE THREAD

**Issue by claudiopro** [emeraldion/emerails#31]
# Summary
All relationship methods accept a `'join'` clause to save additional `SELECT` queries for associated models. See #24 for the rationale.

# Proposed syntax
## `has_one`
```php
$car->has_one(LicensePlate::class, array(
  'join' => Color::class
));
```

## `has_many`
```php
$car->has_many(Part::class, array(
  'join' => Brand::class
));
```

## `belongs_to`
```php
$license_plate->belongs_to(Car::class, array(
  'join' => Make::class
));
```

## `has_and_belongs_to_many`
```php
$shopping_list->has_and_belongs_to_many(Product::class, array(
  'join' => Brand::class
));
```

# Issues

The `has_one` method has the signature `has_one($class_or_table_name, $strict = false)` and adding the `$params` argument in the second slot would require a breaking change to callers. I am pretty sure this is API is seldom, if ever, used.

# Notes

The `has_many` method should already meet the expected behavior as it uses `find_all` under the hood.

# Progress
- [x] has_one
- [x] has_many
- [x] belongs_to
- [x] has_and_belongs_to_many

**Comment by claudiopro** [comment]
Implemented join in `has_and_belongs_to_many` with c221f985ee700f716344660c5909f089669fde75

**Comment by claudiopro** [comment]
This should be done with b54142a59054cd5d752f02204cb2d23a0dad285f
